File: python/algorithms/merge_sort_top_down.py
#!/usr/bin/env python


def merge_sort_top_down(iterable):
  """Sorts the iterable.

  Merge sort is a divide and conquer, comparison sorting algorithm.
  Conceptually, the steps are:
  * Divide the collection into n smaller collections, each containing one item,
    which are sorted by definition.
  * Repeatedly merge the smaller collections into larger sorted collections
    until there is only one sorted collection remaining.

  * O(n log n) time complexity
  * O(n) space complexity
  * stable
  * not adaptive

  Args:
    iterable: A collection that is iterable.
  Raises:
    TypeError: If iterable is not iterable.
  """
  try:
    _ = iter(iterable)
  except TypeError:
    raise TypeError('\'{}\' object is not iterable'.format(
        type(iterable).__name__))

  def _merge(iterable, temp, low, middle, high):
    for i in range(low, high + 1):
      temp[i] = iterable[i]
    i = low
    j = low
    k = middle + 1
    while j < middle + 1 and k < high + 1:
      if temp[j] <= temp[k]:
        iterable[i] = temp[j]
        j += 1
      else:
        iterable[i] = temp[k]
        k += 1
      i += 1
    while j < middle + 1:
      iterable[i] = temp[j]
      i += 1
      j += 1

  def _merge_sort_top_down(iterable, temp, low, high):
    if low >= high:
      return
    middle = int(low + (high - low) / 2)
    _merge_sort_top_down(iterable, temp, low, middle)
    _merge_sort_top_down(iterable, temp, middle + 1, high)
    _merge(iterable, temp, low, middle, high)

  length = len(iterable)
  temp = [None] * length
  _merge_sort_top_down(iterable, temp, 0, length - 1)

File: c/algorithms/merge_sort_top_down.c
#include "merge_sort_top_down.h"

#include <stdlib.h>

static void merge(int array[],
                  int temp[],
                  size_t low,
                  size_t middle,
                  size_t high) {
  // Copy both haves of array to temp.
  for (size_t i = low; i < high + 1; ++i) {
    temp[i] = array[i];
  }

  // Copy the next largest element from temp to array.
  size_t i = low;
  size_t j = low;
  size_t k = middle + 1;
  while (j < middle + 1 && k < high + 1) {
    if (temp[j] < temp[k]) {
      array[i] = temp[j];
      ++j;
    } else {
      array[i] = temp[k];
      ++k;
    }
    ++i;
  }

  // Copy the remaining elements from temp to array.
  while (j < middle + 1) {
    array[i] = temp[j];
    ++i;
    ++j;
  }
}

static void merge_sort_top_down_internal(int array[],
                                         int temp[],
                                         size_t low,
                                         size_t high) {
  if (low >= high) {
    return;
  }
  size_t middle = low + (high - low) / 2;
  merge_sort_top_down_internal(array, temp, low, middle);
  merge_sort_top_down_internal(array, temp, middle + 1, high);
  merge(array, temp, low, middle, high);
}

void merge_sort_top_down(int array[], size_t length) {
  if (length == 0) {
    return;
  }

  int* temp = malloc(length * sizeof(int));
  merge_sort_top_down_internal(array, temp, 0, length - 1);
  free(temp);
}
